File: playground/bench/search.bench.ts
/**
 * Vitest benchmark: FuzzySearch (nlptools) vs Fuse.js
 *
 * Compares result quality and search speed across different collection sizes.
 */

import { describe, bench } from "vitest";
import Fuse from "fuse.js";
import { FuzzySearch, findBestMatch } from "@nlptools/distance";

const BENCH_CONFIG = {
  iterations: 1000,
  time: 1000,
};

// ============================================================================
// Test data
// ============================================================================

const BOOKS = [
  { title: "Old Man's War", author: "John Scalzi" },
  { title: "The Lock Artist", author: "Steve Hamilton" },
  { title: "The Great Gatsby", author: "F. Scott Fitzgerald" },
  { title: "Great Expectations", author: "Charles Dickens" },
  { title: "The Hunger Games", author: "Suzanne Collins" },
  { title: "Harry Potter", author: "J.K. Rowling" },
  { title: "To Kill a Mockingbird", author: "Harper Lee" },
  { title: "Pride and Prejudice", author: "Jane Austen" },
  { title: "The Catcher in the Rye", author: "J.D. Salinger" },
  { title: "Brave New World", author: "Aldous Huxley" },
  { title: "Lord of the Flies", author: "William Golding" },
  { title: "The Hobbit", author: "J.R.R. Tolkien" },
  { title: "Fahrenheit 451", author: "Ray Bradbury" },
  { title: "Moby Dick", author: "Herman Melville" },
  { title: "War and Peace", author: "Leo Tolstoy" },
  { title: "Crime and Punishment", author: "Fyodor Dostoevsky" },
  { title: "The Odyssey", author: "Homer" },
  { title: "Don Quixote", author: "Miguel de Cervantes" },
  { title: "1984", author: "George Orwell" },
  { title: "Animal Farm", author: "George Orwell" },
];

const STRING_LIST = [
  "apple",
  "banana",
  "cherry",
  "date",
  "elderberry",
  "fig",
  "grape",
  "honeydew",
  "kiwi",
  "lemon",
  "mango",
  "nectarine",
  "orange",
  "pear",
  "quince",
  "raspberry",
  "strawberry",
  "tangerine",
  "watermelon",
  "blueberry",
];

const QUERIES = [
  "old man",
  "grate gatsbi",
  "hary poter",
  "jane austn",
  "george orwel",
  "frnkenstein",
];

// ============================================================================
// Correctness: compare top results
// ============================================================================

function assertTopMatch(label: string, actual: string, expected: string) {
  if (actual !== expected) {
    throw new Error(`${label}: top match "${actual}" !== expected "${expected}"`);
  }
}

// Run correctness checks once, print results
console.log("\n=== Correctness: FuzzySearch vs Fuse.js ===\n");

{
  // Object array: title key
  const fuse = new Fuse(BOOKS, { keys: ["title"], includeScore: true });
  const nlptools = new FuzzySearch(BOOKS, { keys: [{ name: "title" }], algorithm: "levenshtein" });

  console.log("Object array (title key, levenshtein):");
  for (const q of QUERIES) {
    const fuseResults = fuse.search(q);
    const nlptoolsResults = nlptools.search(q);
    if (fuseResults.length === 0 && nlptoolsResults.length === 0) {
      console.log(`  "${q}" → no match`);
      continue;
    }
    const fuseTop = fuseResults.length > 0 ? fuseResults[0].item.title : "(none)";
    const fuseScore =
      fuseResults.length > 0 ? ((fuseResults[0] as any).score?.toFixed(3) ?? "N/A") : "N/A";
    const nlptoolsTop =
      nlptoolsResults.length > 0 ? (nlptoolsResults[0].item as any).title : "(none)";
    const nlptoolsScore = nlptoolsResults.length > 0 ? nlptoolsResults[0].score.toFixed(3) : "N/A";
    console.log(
      `  "${q}" → Fuse: "${fuseTop}" (${fuseScore}), NLPTools: "${nlptoolsTop}" (${nlptoolsScore})`,
    );
  }

  // String array: fruit search
  const fuseStr = new Fuse(STRING_LIST, { includeScore: true });
  const nlptoolsStr = new FuzzySearch(STRING_LIST, { algorithm: "levenshtein" });
  const fruitQueries = ["aple", "bannana", "strwbrry", "mngo", "bluberry"];

  console.log("\nString array (levenshtein):");
  for (const q of fruitQueries) {
    const fuseResults = fuseStr.search(q);
    const nlptoolsResults = nlptoolsStr.search(q);
    if (fuseResults.length === 0 && nlptoolsResults.length === 0) continue;
    const fuseTop = fuseResults[0].item;
    const nlptoolsTop = nlptoolsResults[0].item as string;
    assertTopMatch(q, nlptoolsTop, fuseTop);
    console.log(`  "${q}" → "${fuseTop}" (${nlptoolsResults[0].score.toFixed(3)})`);
  }

  // findBestMatch
  const best = findBestMatch("hary poter", BOOKS, {
    keys: [
      { name: "title", weight: 0.7 },
      { name: "author", weight: 0.3 },
    ],
  });
  if (!best) throw new Error("findBestMatch returned null");
  console.log(
    `\nfindBestMatch("hary poter") → "${(best.item as any).title}" (${best.score.toFixed(3)})`,
  );

  // Weighted multi-key with details
  const detailed = new FuzzySearch(BOOKS, {
    keys: [
      { name: "title", weight: 0.7 },
      { name: "author", weight: 0.3 },
    ],
    algorithm: "cosine",
    includeMatchDetails: true,
  });
  const results = detailed.search("george orwel");
  if (results.length === 0) throw new Error("No results for 'george orwel'");
  const top = results[0] as any;
  console.log(
    `weighted cosine("george orwel") → "${top.item.title}" (${top.score.toFixed(3)}) [title=${top.matches.title.toFixed(3)}, author=${top.matches.author.toFixed(3)}]`,
  );

  console.log("\n=== All correctness checks passed ===\n");
}

describe("Correctness: FuzzySearch vs Fuse.js", () => {
  bench(
    "Object array: title key",
    () => {
      const fuse = new Fuse(BOOKS, { keys: ["title"] });
      const nlptools = new FuzzySearch(BOOKS, {
        keys: [{ name: "title" }],
        algorithm: "levenshtein",
      });
      for (const q of QUERIES) {
        const fuseResults = fuse.search(q);
        const nlptoolsResults = nlptools.search(q);
        if (fuseResults.length === 0 && nlptoolsResults.length === 0) continue;
        // Only assert when both have results; Fuse.js threshold may filter more aggressively
        if (fuseResults.length > 0 && nlptoolsResults.length > 0) {
          assertTopMatch(q, (nlptoolsResults[0].item as any).title, fuseResults[0].item.title);
        }
      }
    },
    BENCH_CONFIG,
  );

  bench(
    "String array: fruit search",
    () => {
      const fuse = new Fuse(STRING_LIST);
      const nlptools = new FuzzySearch(STRING_LIST, { algorithm: "levenshtein" });
      const fruitQueries = ["aple", "bannana", "strwbrry", "mngo", "bluberry"];
      for (const q of fruitQueries) {
        const fuseResults = fuse.search(q);
        const nlptoolsResults = nlptools.search(q);
        if (fuseResults.length === 0 && nlptoolsResults.length === 0) continue;
        assertTopMatch(q, nlptoolsResults[0].item as string, fuseResults[0].item);
      }
    },
    BENCH_CONFIG,
  );

  bench(
    "findBestMatch convenience function",
    () => {
      const result = findBestMatch("hary poter", BOOKS, {
        keys: [
          { name: "title", weight: 0.7 },
          { name: "author", weight: 0.3 },
        ],
      });
      if (!result) throw new Error("findBestMatch returned null");
      if ((result.item as any).title !== "Harry Potter") {
        throw new Error(
          `findBestMatch: got "${(result.item as any).title}", expected "Harry Potter"`,
        );
      }
    },
    BENCH_CONFIG,
  );

  bench(
    "Weighted multi-key search",
    () => {
      const nlptools = new FuzzySearch(BOOKS, {
        keys: [
          { name: "title", weight: 0.7 },
          { name: "author", weight: 0.3 },
        ],
        algorithm: "cosine",
        includeMatchDetails: true,
      });
      const results = nlptools.search("george orwel");
      if (results.length === 0) throw new Error("No results for 'george orwel'");
    },
    BENCH_CONFIG,
  );
});

// ============================================================================
// Performance: FuzzySearch vs Fuse.js
// ============================================================================

describe("Performance: String array (20 items)", () => {
  let fuse: Fuse<string>;
  let nlptools: FuzzySearch<string>;

  bench(
    "setup: Fuse.js",
    () => {
      fuse = new Fuse(STRING_LIST);
    },
    BENCH_CONFIG,
  );

  bench(
    "setup: NLPTools FuzzySearch",
    () => {
      nlptools = new FuzzySearch(STRING_LIST);
    },
    BENCH_CONFIG,
  );

  bench(
    "search: Fuse.js x 6 queries",
    () => {
      for (const q of ["aple", "bannana", "strwbrry", "mngo", "bluberry", "grap"]) {
        fuse.search(q);
      }
    },
    BENCH_CONFIG,
  );

  bench(
    "search: NLPTools x 6 queries",
    () => {
      for (const q of ["aple", "bannana", "strwbrry", "mngo", "bluberry", "grap"]) {
        nlptools.search(q);
      }
    },
    BENCH_CONFIG,
  );
});

describe("Performance: Object array (20 items, single key)", () => {
  let fuse: Fuse<any>;
  let nlptools: FuzzySearch<any>;

  bench(
    "setup: Fuse.js",
    () => {
      fuse = new Fuse(BOOKS, { keys: ["title"] });
    },
    BENCH_CONFIG,
  );

  bench(
    "setup: NLPTools FuzzySearch",
    () => {
      nlptools = new FuzzySearch(BOOKS, { keys: [{ name: "title" }] });
    },
    BENCH_CONFIG,
  );

  bench(
    "search: Fuse.js x 6 queries",
    () => {
      for (const q of QUERIES) fuse.search(q);
    },
    BENCH_CONFIG,
  );

  bench(
    "search: NLPTools x 6 queries",
    () => {
      for (const q of QUERIES) nlptools.search(q);
    },
    BENCH_CONFIG,
  );
});

describe("Performance: Object array (20 items, multi-key weighted)", () => {
  let fuse: Fuse<any>;
  let nlptools: FuzzySearch<any>;

  bench(
    "setup: Fuse.js",
    () => {
      fuse = new Fuse(BOOKS, {
        keys: [
          { name: "title", weight: 0.7 },
          { name: "author", weight: 0.3 },
        ],
      });
    },
    BENCH_CONFIG,
  );

  bench(
    "setup: NLPTools FuzzySearch",
    () => {
      nlptools = new FuzzySearch(BOOKS, {
        keys: [
          { name: "title", weight: 0.7 },
          { name: "author", weight: 0.3 },
        ],
      });
    },
    BENCH_CONFIG,
  );

  bench(
    "search: Fuse.js x 6 queries",
    () => {
      for (const q of QUERIES) fuse.search(q);
    },
    BENCH_CONFIG,
  );

  bench(
    "search: NLPTools x 6 queries",
    () => {
      for (const q of QUERIES) nlptools.search(q);
    },
    BENCH_CONFIG,
  );
});

describe("Performance: findBestMatch one-shot", () => {
  bench(
    "findBestMatch x 6 queries",
    () => {
      for (const q of QUERIES) findBestMatch(q, BOOKS, { keys: [{ name: "title" }] });
    },
    BENCH_CONFIG,
  );
});

describe("Performance: Different algorithms", () => {
  bench(
    "levenshtein x 6 queries",
    () => {
      const s = new FuzzySearch(BOOKS, { keys: [{ name: "title" }], algorithm: "levenshtein" });
      for (const q of QUERIES) s.search(q);
    },
    BENCH_CONFIG,
  );

  bench(
    "cosine x 6 queries",
    () => {
      const s = new FuzzySearch(BOOKS, { keys: [{ name: "title" }], algorithm: "cosine" });
      for (const q of QUERIES) s.search(q);
    },
    BENCH_CONFIG,
  );

  bench(
    "jaccard x 6 queries",
    () => {
      const s = new FuzzySearch(BOOKS, { keys: [{ name: "title" }], algorithm: "jaccard" });
      for (const q of QUERIES) s.search(q);
    },
    BENCH_CONFIG,
  );

  bench(
    "sorensen x 6 queries",
    () => {
      const s = new FuzzySearch(BOOKS, { keys: [{ name: "title" }], algorithm: "sorensen" });
      for (const q of QUERIES) s.search(q);
    },
    BENCH_CONFIG,
  );

  bench(
    "jaccardNgram x 6 queries",
    () => {
      const s = new FuzzySearch(BOOKS, { keys: [{ name: "title" }], algorithm: "jaccardNgram" });
      for (const q of QUERIES) s.search(q);
    },
    BENCH_CONFIG,
  );
});

File: packages/distance-wasm/tests/search_bench.rs
#![cfg(test)]

use distance_wasm::*;
use distance_wasm::search::*;
use std::time::Instant;

const BOOKS: &[&str] = &[
    "Old Man's War",
    "The Lock Artist",
    "The Great Gatsby",
    "Great Expectations",
    "The Hunger Games",
    "Harry Potter",
    "To Kill a Mockingbird",
    "Pride and Prejudice",
    "The Catcher in the Rye",
    "Brave New World",
    "Lord of the Flies",
    "The Hobbit",
    "Fahrenheit 451",
    "Moby Dick",
    "War and Peace",
    "Crime and Punishment",
    "The Odyssey",
    "Don Quixote",
    "1984",
    "Animal Farm",
];

const STRING_LIST: &[&str] = &[
    "apple",
    "banana",
    "cherry",
    "date",
    "elderberry",
    "fig",
    "grape",
    "honeydew",
    "kiwi",
    "lemon",
    "mango",
    "nectarine",
    "orange",
    "pear",
    "quince",
    "raspberry",
    "strawberry",
    "tangerine",
    "watermelon",
    "blueberry",
];

const SEARCH_QUERIES: &[&str] = &[
    "old man",
    "grate gatsbi",
    "hary poter",
    "jane austn",
    "george orwel",
    "frnkenstein",
];

const FRUIT_QUERIES: &[&str] = &["aple", "bannana", "strwbrry", "mngo", "bluberry"];

const ITERATIONS: u64 = 1000;

fn bench_search(
    label: &str,
    items: &[&str],
    queries: &[&str],
    algo: fn(&str, &str) -> f64,
) {
    let start = Instant::now();
    for _ in 0..ITERATIONS {
        for q in queries {
            let q_lower = q.to_lowercase();
            for item in items {
                let item_lower = item.to_lowercase();
                let _ = algo(&q_lower, &item_lower);
            }
        }
    }
    let total = ITERATIONS as f64 * queries.len() as f64;
    let avg_us = start.elapsed().as_micros() as f64 / total;
    println!("  {:<40} | {:>10.2} us/query", label, avg_us);
}

#[test]
fn bench_search_algorithms() {
    println!("\n=== Search: Different Algorithms (20 items, 6 queries) ===");
    bench_search(
        "levenshtein x 6 queries",
        STRING_LIST,
        SEARCH_QUERIES,
        levenshtein_normalized,
    );
    bench_search("cosine x 6 queries", STRING_LIST, SEARCH_QUERIES, cosine);
    bench_search("jaccard x 6 queries", STRING_LIST, SEARCH_QUERIES, jaccard);
    bench_search("sorensen x 6 queries", STRING_LIST, SEARCH_QUERIES, sorensen);
    bench_search(
        "jaccard_bigram x 6 queries",
        STRING_LIST,
        SEARCH_QUERIES,
        jaccard_bigram,
    );
}

#[test]
fn bench_search_object_array() {
    println!("\n=== Search: Object Array (20 books, title key) ===");
    bench_search(
        "levenshtein x 6 queries",
        BOOKS,
        SEARCH_QUERIES,
        levenshtein_normalized,
    );
    bench_search("cosine x 6 queries", BOOKS, SEARCH_QUERIES, cosine);
}

#[test]
fn bench_search_fuzzy() {
    println!("\n=== Search: Fruit Fuzzy Matching (20 items, 5 queries) ===");
    bench_search(
        "levenshtein x 5 queries",
        STRING_LIST,
        FRUIT_QUERIES,
        levenshtein_normalized,
    );
}

#[test]
fn bench_fuzzy_search_class() {
    let string_items: Vec<String> = STRING_LIST.iter().map(|s| s.to_string()).collect();
    let book_items: Vec<String> = BOOKS.iter().map(|s| s.to_string()).collect();

    println!("\n=== FuzzySearch class (20 items) ===");

    // String array search
    let fs = FuzzySearch::new(string_items.clone(), Algorithm::Levenshtein, 0.3, false);
    let start = Instant::now();
    for _ in 0..ITERATIONS {
        for q in SEARCH_QUERIES {
            let _ = fs.search(q, None);
        }
    }
    let total = ITERATIONS as f64 * SEARCH_QUERIES.len() as f64;
    let avg_us = start.elapsed().as_micros() as f64 / total;
    println!("  {:<40} | {:>10.2} us/query", "string array (levenshtein)", avg_us);

    // Object array (title key) search
    let fs2 = FuzzySearch::new(book_items.clone(), Algorithm::Levenshtein, 0.3, false);
    let start = Instant::now();
    for _ in 0..ITERATIONS {
        for q in SEARCH_QUERIES {
            let _ = fs2.search(q, None);
        }
    }
    let total = ITERATIONS as f64 * SEARCH_QUERIES.len() as f64;
    let avg_us = start.elapsed().as_micros() as f64 / total;
    println!("  {:<40} | {:>10.2} us/query", "object/title (levenshtein)", avg_us);

    // Multi-key weighted search (title=0.7, author=0.3)
    let titles: Vec<String> = BOOKS.iter().map(|s| s.to_string()).collect();
    let authors: Vec<String> = vec![
        "John Scalzi", "Steve Hamilton", "F. Scott Fitzgerald", "Charles Dickens",
        "Suzanne Collins", "J.K. Rowling", "Harper Lee", "Jane Austen",
        "J.D. Salinger", "Aldous Huxley", "William Golding", "J.R.R. Tolkien",
        "Ray Bradbury", "Herman Melville", "Leo Tolstoy", "Fyodor Dostoevsky",
        "Homer", "Miguel de Cervantes", "George Orwell", "George Orwell",
    ].into_iter().map(|s| s.to_string()).collect();
    let mut key_values = Vec::new();
    for i in 0..BOOKS.len() {
        key_values.push(titles[i].clone());
        key_values.push(authors[i].clone());
    }
    let mks = MultiKeyFuzzySearch::new(key_values, 2, vec![0.7, 0.3], Algorithm::Levenshtein, 0.3, false);
    let start = Instant::now();
    for _ in 0..ITERATIONS {
        for q in SEARCH_QUERIES {
            let _ = mks.search(q, None);
        }
    }
    let total = ITERATIONS as f64 * SEARCH_QUERIES.len() as f64;
    let avg_us = start.elapsed().as_micros() as f64 / total;
    println!("  {:<40} | {:>10.2} us/query", "object/title+author (weighted)", avg_us);
}
